### ISSUE TITLE
Questions about custom multi-tenancy setup

### ISSUE THREAD

**Issue by VitoRomaniello** [Finbuckle/Finbuckle.MultiTenant#133]
On my database SQL, using NHibernate, i have stored Tenant table one to many to User table and i have tenants.json with settings for each tenant like Id, Name, Theme etc...

When a user logs in to the web site I have to check in the database which tenant corresponds to.
Then, during the user session I would use, for example, TenantContext with specific tenant data in all the pages .cshtml i have needed.

Which of the [MultiTenant Strategies](https://www.finbuckle.com/MultiTenant/Docs/Strategies) should I use or which way should i follow?
Thank you very much!

**Comment by AndrewTriesToCode** [comment]
hi @VitoRomaniello 

Let me see if I understand your setup: you don't really know the tenant until they have logged in, right? In the [SharedLoginSample](https://github.com/Finbuckle/Finbuckle.MultiTenant/tree/master/samples/SharedLoginSample) I did something similar--I used the domain part of the email to determine the tenant, then I changed the `MultiTenantContext.TenantInfo` to reflect the detected tenant (I had to use reflection for this... I will make it easier in the next release). Then I signed in and redirected the user. In this case I was using Route strategy so I redirected them to an appropriate route for their tenant.

So in short this might work for you:
1) user submits log in info, you look them up in the table to see the tenant (don't actually complete the sign in yet)
2) change `MultiTenantContext.TenantInfo` to the detected tenant (see the [sample](https://github.com/Finbuckle/Finbuckle.MultiTenant/tree/master/samples/SharedLoginSample) to see how I did this.)
3) sign in the user (the cookie value will be adjusted for the tenant)
4) redirect to a route appropriate for the tenant

Good luck and let me know how it goes!

**Comment by VitoRomaniello** [comment]
Hi Andrew, thanks for the suggestion.  Yes i don't know the Tenant until an user has logged in the site.
I only have to check the user's tenant on database and put tenant data on cookie value.
But i don't understand how to use these:
```
var tenantInfo = await store.TryGetByIdentifierAsync(identifier);
var multiTenantContext = HttpContext.GetMultiTenantContext();
multiTenantContext.GetType().GetProperty("TenantInfo").SetValue(multiTenantContext, tenantInfo);
```
What is the identifier and how can i change the MultiTenantContext.TenantInfo in my case?

Can you give me please an example of the right way for my solution?
Thank you so much for the availability.

**Comment by AndrewTriesToCode** [comment]
Hi again @VitoRomaniello 
In the code above for the identifier I was using the the @_example.com_ part of the email address. With that as the identifier I could get the full `TenantInfo` from the `TenantStore`. The last line is just using reflection to set the new `TenantInfo` on the apps current `MultiTenantContext`. Ideally I could have just said `multiTenantContext.TenantInfo = tenantInfo` but `multiTenantContext.TenantInfo` is a read-only property so in my example I cheated and used reflection to set it. (In the next release I will make it read-write).

In your case you will actually need to query your database with the users full email to get the tenant they belong to. You will ultimately need to get a `TenantInfo` object and set it on the current `MultiTenantContext`. Only then can you actually sign them in (the sign in sets cookies and the `TenantInfo` needs to be in place for this to work).

If you post your code or link me to a GitHub for it I can take a closer look.

**Comment by VitoRomaniello** [comment]
Hi Andrew, first of all thanks for the support.
In Startup.cs i put only this code (i don't need route strategy or in memory store for now):

```
services.AddMultiTenant()
    //.WithRouteStrategy(ConfigRoutes)
    //.WithInMemoryStore(Configuration.GetSection("Finbuckle:MultiTenant:InMemoryStore"))
    .WithPerTenantOptions<CookieAuthenticationOptions>((options, tenantInfo) =>
        {
           options.Cookie.Name = $"{tenantInfo.Id}_{options.Cookie.Name}";
           options.Cookie.Path = $"/{tenantInfo.Identifier}";
        });
```

Then in my SharedLogin.cs:
```

 if (ModelState.IsValid)
{
//Get tenant information from query on NHibernate database with mediator pattern
var tenantInfo = await mediator.Send(request);

if (tenantInfo == null)
{
    ModelState.TryAddModelError("", "Tenant not found.");
    goto SkipToEnd;
}

// I don't understand how to do the following code:
var multiTenantContext = HttpContext.GetMultiTenantContext();
multiTenantContext.GetType().GetProperty("TenantInfo").SetValue(multiTenantContext, tenantInfo);
var orignalServiceProvider = HttpContext.RequestServices;
HttpContext.RequestServices = orignalServiceProvider.CreateScope().ServiceProvider;

return RedirectToAction("Index", "Home", new { __tenant__ = tenantInfo.Identifier });
}
```

**Comment by AndrewTriesToCode** [comment]
Hi @VitoRomaniello , sorry I couldn't reply sooner. I can see you are able to get the `TenantInfo` from your mediator. The part below that appears correct-- its using the `TenantInfo` object you got from the mediator and assigning it as the current tenant.

You don't appear to have the code that does the actual sign-in (which sets the sign in cookie for the tenant). Are you using Identity or just handling your own sign-in? How are you checking the uesrname/password and setting a sign-in cookie?

In my code I was using ASP.NET Core Identity:
```
 var signInManager = HttpContext.RequestServices 
     .GetRequiredService<SignInManager<MultiTenantIdentityUser>>();
 var result = await signInManager.PasswordSignInAsync(model.Email, model.Password, false, false);
```

**Comment by VitoRomaniello** [comment]
Thank you so much @achandlerwhite.
